internal/controller/dnsrecord_healthchecks.go

```diff
@@ -0,0 +1,242 @@
+package controller
+
+import (
+	"context"
+	"crypto/md5"
+	"fmt"
+	"io"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/net"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	externaldns "sigs.k8s.io/external-dns/endpoint"
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+	"github.com/kuadrant/dns-operator/internal/provider"
+)
+
+// healthChecksConfig represents the user configuration for the health checks
+type healthChecksConfig struct {
+	Endpoint         string
+	Port             *int64
+	FailureThreshold *int64
+	Protocol         *provider.HealthCheckProtocol
+}
+
+func (r *DNSRecordReconciler) ReconcileHealthChecks(ctx context.Context, dnsRecord *v1alpha1.DNSRecord) error {
+	var results []provider.HealthCheckResult
+	var err error
+
+	// Get the configuration for the health checks. If no configuration is
+	// set, ensure that the health checks are deleted
+	config := getHealthChecksConfig(dnsRecord)
+
+	log.Log.Info("DNS Record healthcheck status", "status", dnsRecord.Status.HealthCheck)
+	if config == nil {
+		log.Log.V(3).Info("No health check for DNSPolicy, ensuring deletion")
+		results, err = r.reconcileHealthCheckDeletion(ctx, dnsRecord)
```

Comment:
having a reconcile for deletion and a reconcile for adding / updating feels a bit odd? Why can't it just be done in the one reconcile?

Reply:
this was brought in from the old MGC implementation, I can change it

Reply:
well doesn't block merging, but just seems a bit cumbersome